### PR TITLE
Fix crasher caused by feature count (fixes #16470)

### PR DESCRIPTION
**Pull request by nirvn**
## Description
This PR disables the progress dialog for feature count in the QgsDefaultVectorLayerLegend::createLayerTreeModelLegendNodes() function to fix at least one identified, reproducible crasher (https://issues.qgis.org/issues/16470). 

The problem with the progress dialog is that it calls QCoreApplication::processEvents() , which in turn causes QGIS-specific signals to be emitted earlier than expected.

It's the second time this year I stumbled on an issue resulting from this progress dialog (the first one being an unwarranted changed of projection upon project load).

@nyalldawson , @wonder-sk , @m-kuhn , IMHO, this PR should be applied. The one downside is that we introduce a UI freeze when a large dataset is being counted, but, `crash > freeze`.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by nirvn** [comment]
BTW, I wouldn't be surprised this progress dialog causes other unwanted side effects, which we simply aren't aware of.

**Comment by nyalldawson** [comment]
Is the dialog used anywhere else? If not, there's lots of code which can be removed from countSymbolFeatures.

I think ideally the bool showProgress should be replaced by a `QgsFeedback *feedback = nullptr` argument. Then the cancelation support can be changed over from the progress dialog to the feedback object. This would help if ever in future the count was moved to a background task.

This also helps with https://github.com/qgis/qgis3.0_api/issues/66

**Comment by nirvn** [comment]
I'm in favor of removing the code altogether.

@wonder-sk , @m-kuhn , speak now or forever remain silent ... 😉

**Comment by nyalldawson** [comment]
> I'm in favor of removing the code altogether.

I'd still leave it, just flip it over to the QgsFeedback. All these methods like this (minimum/maximumValue, uniqueValues, etc) should ideally all take QgsFeedback and support cancelation.

**Comment by nirvn** [comment]
@nyalldawson , done.

**Comment by nyalldawson** [comment]
Looks good to me. Let's leave it open for a bit longer for more feedback.

**Comment by m-kuhn** [comment]
`processEvents()` has caused us troubles (read: crashes) in the past already.

I'm not completely decided if we should just discourage or actively force that no new code using this is introduced (with something like spellcheck).

**Comment by m-kuhn** [comment]
Meanwhile, how is this code intended to be called?

* If it's from the main thread, is a user even able to make use of the feedback object and send a cancel notification (and introduce `processEvents()` or similar measures at a different point but with the same risks).
* If it's from a background thread, isn't this just as dangerous (anyone can remove the layer while features are being counted on it).

In short, I think methods that live directly in QgsVectorLayer should always be called synchroneously while feedback objects only make sense for methods that are called asynchroneously. But maybe I am mistaken.

**Comment by nyalldawson** [comment]
> In short, I think methods that live directly in QgsVectorLayer should always be called synchroneously while feedback objects only make sense for methods that are called asynchroneously. But maybe I am mistaken.

I disagree somewhat - I think we need to leave this up to the caller to decide how to handle this. If we don't add the feedback options here, there's NO option to run these calculations in the background. But if we do add them, then a caller can (carefully) run these calculations in the background, with appropriate hooks to layer removal to cancel when required.

**Comment by m-kuhn** [comment]
> then a caller can (carefully) run these calculations in the background, with appropriate hooks to layer removal to cancel when required.

I do not think this is safely possible.

* Deleting a layer is only one of a couple of problematic scenarios.
* The only signal that is reliably sent when deleting a layer is `destroyed` but too late, some destruction has already run then
* New members might be accessed inside the method in the future. Even if you write "safe" code today, suddenly there might be new race conditions

**Comment by nirvn** [comment]
Arg, got hit by that crash again.

What's the decision on this?

**Comment by m-kuhn** [comment]
I don't mind to apply the quickfix itself.

I am only hesitating to indicate that something is capable of being called async by adding a `QgsFeedback` object, when in fact it should always be called sync. The only clean solution is IMHO to create a `class QgsFeatureCounter : public QgsTask`.

**Comment by nyalldawson** [comment]
> The only clean solution is IMHO to create a class QgsFeatureCounter : public QgsTask

Not necessarily - what about moving them to QgsVectorLayerUtils? then the feedback object can be added without the risk associated with the methods belonging to potentially deletable layers?

**Comment by m-kuhn** [comment]
If this internally takes a copy of all the involved data and drops the ptr to QgsVectorLayer immediately (in blocking setup code always executed on the main thread) ... Then yes.

**Comment by m-kuhn** [comment]
Actually it might even work with a QPointer<QgsVectorLayer> which is checked continuously after each processEvents incocation. But it still feels all like an advanced workaround. IMHO it should be one of

- make it multithreaded
- keep it blocking. no feedback, no processEvents. Bad UX but better than crashes.

**Comment by nyalldawson** [comment]
@m-kuhn I'm still unconvinced. There's plenty of ways the QGIS API can be made to crash, and I don't think it's necessary to block every unsafe use if it also blocks valid use.

E.g. let's say I want to make categorised classification non blocking (which I do). I'd need to add the feedback object to the QgsVectorLayer/provider functions to be able to do this. While this could be abused and used in a crashy way, it also can be used in a totally safe way inside the style widget (where the destructor would block while a request is processing, in turn blocking the layer destruction). But without the feedback object this wouldn't be possible at all.

I'd prefer to have a more permissive API which allows use like this, even if it can be used in an unsafe manner.

Or am I misinterpreting your concerns here?

**Comment by m-kuhn** [comment]
See also #4526 (the only testing done so far is: it compiles for me ;) )

As far as I can see it's not about things that "could be abused and used in a crashy way" but rather about things that are almost "impossible to use in a safe way". While they might work well in 99% of the cases, as soon as `processEvents()` is involved (or things are sent to a background task) things might (and eventually will) happen in an unforeseen order, base data will change or objects be deleted (in events called from the event loop). If it was about a warning which pops up in 1% of the cases, I wouldn't mind too much. If it's about loosing unsaved work because of a crash I'm in favor of playing this extra-safe.

**Comment by m-kuhn** [comment]
> (where the destructor would block while a request is processing, in turn blocking the layer destruction)

Not sure I follow which destructor you mean.
If it's the layer destructor, there's [probably something wrong](http://stackoverflow.com/a/646429), I didn't read the standard but in the past I experienced problems with partially deleted objects when reacting to events in the destructor.
If it's some dialog destructor (or modal dialog) it's probably a bit safer because the user can't really click and do something unexpected.

But let's say the layer was created by a processing alg and it's only there temporary and it's a script that deletes the layer while the features are being counted... The example is totally made up, but something along these lines triggered the crashes which have led to this pull request.

**Comment by nyalldawson** [comment]
> . While they might work well in 99% of the cases, as soon as processEvents() is involved

Oh - my entire conversation here is under the premise that processEvents() is an evil beast from the depths of hades and should be banished from core altogether. So I'm assuming that it's not part of the equation here...

> Not sure I follow which destructor you mean.

The destructor for the layer properties/renderer properties widget will catch this. I.e. style dock listens out for layer about to be removed, if so, it destroys the renderer widget - which catches the running calculation and blocks until it's canceled.

**Comment by nirvn** [comment]
This is superseded by @m-kuhn 's PR #4526 , closing.
